### ISSUE TITLE
[temp.constr.normal] The obscure(contradictory) of the normalization of a concept-id

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/draft#5389]
[temp.constr.normal] p1.4 says
> The normal form of a concept-id `C<A1, A2,..., An>` is the normal form of the [constraint-expression](https://eel.is/c++draft/temp.constr.decl#nt:constraint-expression) of C, **after** substituting `A1, A2,..., An` for C's respective template parameters in the parameter mappings in each atomic constraint[.](https://eel.is/c++draft/temp.constr.normal#1.4.sentence-1) If any such substitution results in an invalid type or expression, the program is ill-formed; no diagnostic is required[.](https://eel.is/c++draft/temp.constr.normal#1.4.sentence-2)  

This sounds like a paradox. "after" means the substitution occurs before the normal form of the constraint-expression of `C`, whereas the substitution requires us to determine the **template parameters** in the parameter mappings in each atomic constraint, which means we should **firstly** determine what atomic constraints are in the normal form of the concept-id.  I think [temp.constr.normal] p1.4 might be changed to
> The normal form of a concept-id `C<A1, A2,..., An>` is the normal form of the constraint-expression of C **by substituting** `A1, A2,..., An` for C's respective template parameters in the parameter mappings in each atomic constraint.  If any such substitution results in an invalid type or expression, the program is ill-formed; no diagnostic is required.  

I think the logic of normalizing an expression is that makes each operand in a logical operation is an atomic constraint.  So, for any expression E, normalization of `E` is reasonable to recursively apply [temp.constr.normal] p1 to `E` so that obtains the atomic constraints, only after which can substitute the template arguments for the template parameter in the parameter mappings apply, as per the definition of parameter mapping.

**Comment by hubert-reinterpretcast** [comment]
> This sounds like a paradox. "after" means the substitution occurs before the normal form of the constraint-expression of `C`

The use of "after" was suboptimal but did not preclude the intended reading that the resulting normal form is arrived at only after said substitution. It has since been removed by https://wg21.link/p2841.

https://wg21.link/temp.constr.normal#1.4 now has:
> The normal form of [ ... ] is the result of substituting, in the normal form [ ... ].

**Comment by jensmaurer** [comment]
It seems that P2841 improved the clarity in this area.  Please open a new issue if you find an issue with the new wording.
